File: android/crosspl/lib/src/main/kotlin/org/elastos/tools/crosspl/CrossBase.kt
package org.elastos.tools.crosspl

import android.util.Log
import org.elastos.tools.crosspl.annotation.CrossClass
import org.elastos.tools.crosspl.annotation.CrossInterface
import kotlin.reflect.KClass
import kotlin.reflect.full.primaryConstructor

@CrossClass
open class CrossBase
    protected constructor(
            className: String,
            var nativeHandle: Long = 0)
{

    init {
        if(nativeHandle == 0L) {
            Log.d(Utils.TAG, "CrossBase CreateNativeObject() className=$className")
            nativeHandle = CreateNativeObject(className)
        }
        Log.i(Utils.TAG, "construct " + toString())
        if(nativeHandle == 0L) {
            throw Utils.CrossPLException("Failed to create native object.")
        }
    }

    open fun finalize() {
        DestroyNativeObject(this.javaClass.name, nativeHandle)
        Log.i(Utils.TAG, "deconstruct " + toString())
    }

    open fun bind() {
        bindPlatformHandle(this)
    }

    open fun unbind() {
        unbindPlatformHandle(this)
    }

    @CrossInterface
    override fun toString(): String {
        return "${this.javaClass.name}{nativeHandle=${nativeHandle}}"
    }

    @CrossInterface
    private external fun bindPlatformHandle(thisObj: CrossBase)

    @CrossInterface
    private external fun unbindPlatformHandle(thisObj: CrossBase)


//    private var nativeHandle: Long = 0
    private companion object {
        init {
            Log.d(Utils.TAG, "Loading libcrosspl.so ...")
            System.loadLibrary("crosspl")
            Log.d(Utils.TAG, "Loaded libcrosspl.so ...")
        }

        @CrossInterface
        private external fun CreateNativeObject(className: String): Long

        @CrossInterface
        private external fun DestroyNativeObject(className: String, nativeHandle: Long)

//        @CrossInterface
//        private external fun BindPlatformObject(obj: CrossBase): Long
//
//        @CrossInterface
//        private external fun UnbindPlatformObject(obj: CrossBase): Long
    }
}

File: ios/crosspl/lib/swift/CrossBase.swift
import Foundation

/* @CrossClass */
@objc open class CrossBase : NSObject {
  public init(className: String? = nil, nativeHandle: Int64 = 0) {
    if className != nil {
      self.classTitle = className!
    } else {
      self.classTitle = String(describing: type(of: self))
    }
    self.nativeHandle = nativeHandle

    if(self.nativeHandle == 0) {
      self.nativeHandle = CrossBase.CreateNativeObject(swiftClassName: self.classTitle)
    }
  }
  
  deinit {
    CrossBase.DestroyNativeObject(swiftClassName: self.classTitle, nativeHandle: self.nativeHandle)
    self.nativeHandle = 0
  }
  
  public func bind() {
    bindPlatformHandle(thisObj: self)
  }
  
  public func unbind() {
    unbindPlatformHandle(thisObj: self)
  }

  @objc public private(set) var nativeHandle: Int64
  private let classTitle: String
  
  /* @CrossNativeInterface */
  @objc private static func CreateNativeObject(swiftClassName: String) -> Int64 {
    return crosspl_Proxy_CrossBase_CreateNativeObject(swiftClassName)
  }
  
  /* @CrossNativeInterface */
  @objc private static func DestroyNativeObject(swiftClassName: String, nativeHandle: Int64) {
    crosspl_Proxy_CrossBase_DestroyNativeObject(swiftClassName, nativeHandle)
  }
  
  /* @CrossNativeInterface */
  @objc private func bindPlatformHandle(thisObj: CrossBase) {
    crosspl_Proxy_CrossBase_bindPlatformHandle(nativeHandle, thisObj)
  }
  
  /* @CrossNativeInterface */
  @objc private func unbindPlatformHandle(thisObj: CrossBase) {
    crosspl_Proxy_CrossBase_unbindPlatformHandle(nativeHandle, thisObj)
  }
}
